### PR TITLE
add dilepton ee task

### PR DESCRIPTION
**Pull request by dsekihat**
add dilepton ee task

### CODE REVIEW

**Review by jgrosseo — DISMISSED** [review]

File: Analysis/Tasks/PWGDQ/dileptonEE.cxx
Comment (aalkin) [inline Analysis/Tasks/PWGDQ/dileptonEE.cxx]: This can be simplified by using Partition<>, so that you do not have to accumulate a vector of iterators. You can declare it as
```
using tracktype = soa::Join<aod::ReducedTracks, aod::ReducedTracksBarrel, aod::ReducedTracksBarrelCov, aod::ReducedTracksBarrelPID>;
Partition<tracktype> posTracks = aod::reducedtrack::charge > 0;
```

Comment (dsekihat) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Hi @aalkin ,
Thank you for your comments.
Since I am still a beginner, please allow me to ask some questions.
First, histograms are filled with each track info.
Then, positive/negative tracks are separated for pairing after track selection.
So, I think track loop is still necessary here.

Comment (aalkin) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Accumulating iterators instead of using tables directly most likely prevents optimizations. You can still use `isSelected` in the following double loop over positive and negative partitions.

Comment (dsekihat) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: I see, thanks. Obviously, I have to learn more.
then,
using tracktype = soa::Join<aod::ReducedTracks, aod::ReducedTracksBarrel, aod::ReducedTracksBarrelCov, aod::ReducedTracksBarrelPID>;
should this be at the top of this macro?

Comment (aalkin) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: This is just a convenience alias, you can put it at the beginning of the file and use `tracktype` also in process() declaration.

Comment (jgrosseo) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: You are updating following these comments before we merge, right?

Comment (dsekihat) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: I am trying to update this task.
But, if this is only for optimization and not crucial for hyperloop, please merge this now.
(many PRs seem stuck somehow, though.)
Then, I will continue to optimizing code.

Comment (iarsene) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Hi Anton,
But is it possible to add tracks to a partition after the selection done in fTrackCut->IsSelected() ? If so, this would be great.
The separation over charges is the simple part. The main point is to have a selection cut which can be declared/defined outside the analysis class. This cut can contain criteria based on many derived variables (defined in the VarManager), which are not table columns. Both the number of variables and their limits can be defined by the user outside the analysis class in principle. How can i do that with Partition? I would like to avoid hardcoding in the task. As you maybe saw, in DQ we have histograms and cuts declared outside the analysis. If this is possible with Partition, would be great. We discussed once in a private channel on mattermost exactly about this and you said this is not possible yet, but maybe things changed in the meantime.
Cheers, Ionut

Comment (aalkin) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Hi Ionut,

Currently - no, but looping over separated tracks and using the selection is still better than accumulating an array of iterators. It also may be better to have a multi-stage analysis - so that first task created a derived table with only selected tracks and subsequent tasks manipulate it in a declarative way. There is a PR in testing that will allow to use Configurables in filters and partitions. I understand, that currently the complete set of complicated cuts and selections cannot be implemented this way. However, any selections and partitions that _can_ would be very useful for performance.

The main point here is that iterators should not be treated as "track objects", otherwise there is no benefit of using these data structures at all.

Comment (iarsene) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Hi Anton,
I think i understand what you mean. Note that actually what we do in DQ is to skim the data, so there is a task in front which runs before and produces a small dataset (the Reduced tables). Since these are electrons, we are looking here at about 1 per mille or less of the whole track sample when this task will run in production.
I agree on the performance issues, and we will try to improve our code. But would be also good to quantify these improvements because having user friendly code is also important and somewhere there should be an optimum.
In this particular case i am not sure it will be faster to apply the cut->IsSelected() in the 2-track loop (so applying this cut multiple times on the same track) than using the vector of iterators. But these optimizations can be definitely tried at some point.
In any case, i think that in order to make Partition to be adopted more easily, i think there should be more flexible ways of adding/removing tracks to them. Not sure if possible, but maybe it can be discussed.
Cheers, Ionut

Comment (aalkin) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: In this particular case it may make sense to store the result of isSelected() in a column, join that column to your table, and then you can create a partition both on charge and the selection decision in one line.

Comment (iarsene) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: This sounds like a good suggestion which would be compatible also with the way i am thinking this analysis. I am just not used yet to think in terms of Partition. Thanks for the idea! We even foreseen a special column in the table for these decisions. We just did not thought about using it this way

Comment (dsekihat) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx:126]: Hi Anton and Jan-Fiete,
Thank you for comments. I change my mind.
While I am updating this task, please keep this PR open (please do not merge yet).

@iarsene , Thank you for your help!

Comment (jgrosseo) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Very good. I wait for a push to this PR then.

Comment (dsekihat) [reply Analysis/Tasks/PWGDQ/dileptonEE.cxx]: Hello @jgrosseo and @aalkin ,
Thank you for comments. I implemented same way as https://github.com/AliceO2Group/AliceO2/pull/4291
Many thanks to Ionut!

**Review by jgrosseo — DISMISSED** [review]

**Comment by jgrosseo** [comment]
The clang format check failed

**Comment by dsekihat** [comment]
> The clang format check failed

Thanks for comments. It should be corrected.

**Review by jgrosseo — APPROVED** [review]

**Comment by jgrosseo** [comment]
codechecker CI got aborted twice: c++: internal compiler error: Killed (program cc1plus)

Merging

**Pull request merged**
